ctfcli/utils/challenge.py

```diff
@@ -404,3 +405,228 @@ def lint_challenge(path):
+    r.raise_for_status()
+    hints = r.json()["data"]
+    challenge["hints"] = [{ "content": h["content"], "cost": h["cost"] } if h["cost"] > 0 else h["content"] for h in hints]
+
+    # Add topics
+    r = s.get(f"/api/v1/challenges/{challenge_id}/topics", json=True)
+    r.raise_for_status()
+    topics = r.json()["data"]
+    challenge["topics"] = [t["value"] for t in topics]
+
+    # Add requirements
+    r = s.get(f"/api/v1/challenges/{challenge_id}/requirements", json=True)
+    r.raise_for_status()
+    requirements = (r.json().get("data") or {}).get("prerequisites", [])
+    if len(requirements) > 0:
+        # Prefer challenge names over IDs
+        r = s.get("/api/v1/challenges", json=True)
+        r.raise_for_status()
+        challenges = r.json()["data"]
+        challenge["requirements"] = [c["name"] for c in challenges if c["id"] in requirements]
+
+    return challenge
+
+def is_default(key, value):
+    if key == "connection_info" and value == None:
+        return True
+    if key == "attempts" and value == 0:
+        return True
+    if key == "state" and value == "visible":
+        return True
+    if key == "type" and value == "standard":
+        return True
+    if key in ["tags", "hints", "topics", "requirements"] and value == []:
+        return True
+    return False
+
+def verify_challenge(challenge, ignore=(), verify_files=False, verify_defaults=False, _verify_new_files=True):
+    """
+    Verify that the challenge.yml matches the remote challenge if one exists with the same name
+    """
+    s = generate_session()
+
+    installed_challenges = load_installed_challenges()
+    for c in installed_challenges:
+        if c["name"] == challenge["name"]:
+            challenge_id = c["id"]
+            break
+    else:
+        return
+    
+    remote_challenge = get_challenge_details(challenge_id)
+    for key in remote_challenge:
+        if key in ignore:
+            continue
+        # Special validation needed for files
+        if key == "files":
+            # Get base file name of challenge files
+            local_files = {Path(f).name : f for f in challenge[key]}
+
+            for f in remote_challenge["files"]:
+                # Get base file name
+                f_base = f.split("/")[-1].split('?token=')[0]
+                if f_base not in local_files and _verify_new_files:
+                    raise Exception(f"Remote challenge has file {f_base} that is not present locally")
+                else:
+                    if verify_files:
+                        # Download remote file and compare contents
+                        req = s.get(f)
+                        req.raise_for_status()
+                        remote_file = req.content
+                        local_file = Path(challenge.directory, local_files[f_base]).read_bytes()
+                        if remote_file != local_file:
+                            raise Exception(f"Remote challenge file {f_base} does not match local file")
```

Comment:
Maybe it makes more sense to use [filecmp](https://docs.python.org/3/library/filecmp.html) and write the downloaded file to a tempfile?

Reply:
Just so it's not in memory?